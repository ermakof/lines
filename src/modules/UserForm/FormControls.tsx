/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, memo, useCallback } from 'react';
import { useDispatch } from 'react-redux';

import Button from '@src/components/Button';
import { actions as appActions } from '@src/App/appSlice';
import Select from '@src/components/Select';
import styled from '@emotion/styled';
import { IUserProfile } from '@src/modules/Auth/model/IUserProfile';

const Root = styled.div`
  display: flex;
  width: 100%;
`;

interface IFormControls {
  userLevel?: string;
  userProfile: IUserProfile;
}
const FormControls: FC<IFormControls> = ({ userLevel, userProfile }) => {
  const dispatch = useDispatch();

  const handleReset = useCallback(() => {
    dispatch(
      appActions.updateHitParadeInfo({
        login: userProfile.login,
        ts: userProfile.loginTime,
        score: 0,
      })
    );
    dispatch(appActions.resetApp());
  }, [userProfile]);

  const handleSelectLevel = useCallback(
    (event: React.ChangeEvent<HTMLSelectElement>) => {
      const { value } = event.currentTarget;
      dispatch(
        appActions.updateHitParadeInfo({
          login: userProfile.login,
          ts: userProfile.loginTime,
          score: 0,
        })
      );
      dispatch(appActions.setUserLevel(value));
    },
    [userProfile]
  );

  return (
    <Root role="formControls">
      <Button role="buttonReset" onClick={handleReset} title="Сброс" />
      <Select onSelect={handleSelectLevel} selectedLevel={userLevel} />
    </Root>
  );
};

export default memo(FormControls);
