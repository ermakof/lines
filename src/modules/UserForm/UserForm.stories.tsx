/* eslint-disable */
import React from 'react';
import { Provider } from 'react-redux';
import { ComponentStory, Meta } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs';

import FormControls from '@src/modules/UserForm/FormControls';
import { store } from '@src/store';
import { appSlice } from '@src/App/appSlice';

const payload = {
  gameLevel: '2',
  gameFieldPercentFilled: 10,
  score: 15,
  gameFieldData: [
    0, 1, 0, 0, 0, 1, 1, 1, 1,
    1, 0, 0, 0, 0, 0, 0, 0, 1,
    1, 0, 0, 0, 0, 0, 0, 0, 1,
    1, 0, 0, 0, 0, 0, 0, 0, 1,
    0, 0, 0, 0, 0, 0, 0, 0, 1,
    1, 0, 0, 0, 0, 0, 0, 0, 0,
    1, 1, 1, 1, 0, 0, 0, 0, 0,
    1, 0, 0, 0, 0, 0, 0, 0, 0,
    1, 0, 0, 0, 0, 0, 0, 0, 0,
  ],
};

store.dispatch(appSlice.actions.updateGame(payload));

export default {
  component: FormControls,
  decorators: [withKnobs],
  title: 'Forms/FormControls',
} as Meta;

export const ExampleControls: ComponentStory<typeof FormControls> = (args) => (
  <Provider store={store}>
    <FormControls {...args} />
  </Provider>
);

