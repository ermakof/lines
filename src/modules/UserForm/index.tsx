/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, memo, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import Panel from '@src/layout/Panel';
import { actions as appActions } from '@src/App/appSlice';
import FormControls from '@src/modules/UserForm/FormControls';
import { getApp } from '@src/App/appSaga';
import { getAuth } from '@src/modules/Auth/authSaga';
import styled from '@emotion/styled';
import { actions as authActions } from '@src/modules/Auth/authSlice';
import Button from '@src/components/Button';

const Root = styled(Panel)`
  width: 66%;
  justify-content: left;
`;

const LogoutButton = styled(Button)`
  margin-left: auto;
`;

const UserForm: FC = () => {
  const { userLevel } = useSelector(getApp);
  const { userProfile } = useSelector(getAuth);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(appActions.rehydrate());
  }, []);

  const handleLogout = () => {
    dispatch(authActions.logout(userProfile.loginTime));
  };

  return (
    <Root role="userForm">
      <FormControls userLevel={userLevel} userProfile={userProfile} />
      <LogoutButton role="buttonLogout" onClick={handleLogout} title="Выйти" />
    </Root>
  );
};

export default memo(UserForm);
