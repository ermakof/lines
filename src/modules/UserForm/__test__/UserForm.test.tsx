import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

import UserForm from '@src/modules/UserForm';
import appInitialState from '@src/App/initialState';

const userProfile = { login: 'user', password: 'pwd', loginTime: 123 };

jest.mock('react-redux', () => ({
  __esModule: true,
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

const mockSelectors = (selector: any, store: any) => {
  return selector(store);
};

let mockDispatch: jest.Mock;

describe('UserForm', () => {
  beforeEach(() => {
    (useSelector as jest.Mock).mockImplementation((selector) =>
      mockSelectors(selector, {
        app: appInitialState,
        auth: { userProfile },
      })
    );
    mockDispatch = jest.fn();
    (useDispatch as jest.Mock).mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    (useSelector as jest.Mock).mockClear();
  });

  it('Render form', () => {
    const { asFragment } = render(
      <Router>
        <UserForm />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
    const userForm = screen.getByRole(/userForm/gi);
    expect(userForm).toBeInTheDocument();
    const formControls = screen.getByRole(/^formControls$/gi);
    expect(formControls).toBeInTheDocument();
    const buttonLogout = screen.getByRole(/^buttonLogout$/gi);
    expect(buttonLogout).toBeInTheDocument();
  });

  it('Logout app', async () => {
    render(
      <Router>
        <UserForm />
      </Router>
    );
    userEvent.click(screen.getByRole('buttonLogout'));

    await waitFor(() =>
      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'auth/logout',
        payload: 123,
      })
    );
  });
});
