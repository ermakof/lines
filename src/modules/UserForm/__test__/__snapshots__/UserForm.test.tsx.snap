// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm Render form 1`] = `
<DocumentFragment>
  <div
    class="css-1xojm1q"
    role="userForm"
  >
    <div
      class="css-123spcd"
      role="formControls"
    >
      <button
        class="css-tv5zn1"
        role="buttonReset"
      >
        Сброс
      </button>
      <div
        class="css-1xvr7bb"
      >
        <p>
          Уровень
        </p>
        <select
          name="select"
          role="select"
        >
          <option
            role="option"
            value="1"
          >
            Падаван
          </option>
          <option
            role="option"
            value="2"
          >
            Джедай
          </option>
          <option
            role="option"
            value="3"
          >
            Член Совета Силы
          </option>
        </select>
      </div>
    </div>
    <button
      class="css-tv5zn1"
      role="buttonLogout"
    >
      Выйти
    </button>
  </div>
</DocumentFragment>
`;
